pandas/_libs/parsers.pyx

```diff
@@ -1267,19 +1267,49 @@ cdef class TextReader:
             return self._string_convert(i, start, end, na_filter,
                                         na_hashset)
         elif is_categorical_dtype(dtype):
+            # TODO: I suspect that _categorical_convert could be
+            # optimized when dtype is an instance of CategoricalDtype
             codes, cats, na_count = _categorical_convert(
                 self.parser, i, start, end, na_filter,
                 na_hashset, self.c_encoding)
-            # sort categories and recode if necessary
             cats = Index(cats)
-            if not cats.is_monotonic_increasing:
+
+            # Determine if we should convert inferred string
+            # categories to a specialized type
+            if (isinstance(dtype, CategoricalDtype) and
```

Comment:
I would rather move this entire section to a free function (except for the actual constructor)

maybe
```
cats, dtype = infer_categorical_dtype(cats) # put in pandas.core.dtypes.cast.py
cats = Categorical(cats, codes, dtype=dtype)
```

NONE of this logic should be here